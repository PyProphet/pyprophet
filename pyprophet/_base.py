from dataclasses import dataclass, field
import os
import copy
import sys

from loguru import logger
from .io.util import (
    is_tsv_file,
    is_sqlite_file,
    is_parquet_file,
    is_valid_single_split_parquet_dir,
    is_valid_multi_split_parquet_dir,
)


@dataclass
class BaseIOConfig:
    """
    Base configuration class for I/O-related metadata and common attributes used across algorithms.

    Attributes:
        infile (str): Path to the input file (e.g., .osw, .parquet).
        outfile (str): Path to the output file to be written.
        file_type (str): Type of the input file (e.g., 'osw', 'parquet', 'parquet_split', 'tsv').
        subsample_ratio (float): Subsampling ratio for large data.
        level (str): Scoring level (e.g., 'ms1', 'ms2', 'transition', 'alignment').
        context (str): Context or mode in which the reader/writer operates
                       (e.g., 'score_learn', 'score_apply', 'ipf', 'level_context').
        prefix (str): Automatically derived from outfile (e.g., 'results/output' for 'results/output.osw').
    """

    infile: str
    outfile: str
    file_type: str = field(init=False)
    subsample_ratio: float
    level: str
    context: str
    prefix: str = field(init=False)

    def __post_init__(self):
        """
        Initialize the file_type and prefix attributes based on the input file.
        """

        infile = self.infile

        if is_sqlite_file(infile):
            if infile.endswith(".osw"):
                self.file_type = "osw"
            elif infile.endswith(".sqmass") or infile.endswith(".sqMass"):
                self.file_type = "sqmass"
        elif is_parquet_file(infile):
            self.file_type = "parquet"
        elif is_valid_single_split_parquet_dir(infile):
            self.file_type = "parquet_split"
        elif is_valid_multi_split_parquet_dir(infile):
            self.file_type = "parquet_split_multi"
        elif is_tsv_file(infile):
            self.file_type = "tsv"
        else:
            logger.critical(
                f"Failed to infer file type for: {infile}. Supported formats are: "
                ".osw, .sqmass/.sqMass, .parquet, split parquet directories (.oswpq/oswpqd), or .tsv files.\n"
                f"  - is_sqlite_file: {is_sqlite_file(infile)}\n"
                f"  - endswith .osw: {infile.endswith('.osw')}\n"
                f"  - endswith .sqmass/.sqMass: {infile.endswith('.sqmass') or infile.endswith('.sqMass')}\n"
                f"  - is_parquet_file: {is_parquet_file(infile)}\n"
                f"  - is_valid_single_split_parquet_dir: {is_valid_single_split_parquet_dir(infile)}\n"
                f"  - is_valid_multi_split_parquet_dir: {is_valid_multi_split_parquet_dir(infile)}\n"
                f"  - is_tsv_file: {is_tsv_file(infile)}"
            )

            sys.exit(1)

        self.prefix = os.path.splitext(
            self.outfile
        )[
            0
        ]  # TODO: use pathlib instead to avoid potential cases where the outfile is a directory (split_parquet) file.oswpq/, which would result the prefix in being file.oswpq instead of file

    def __str__(self):
        return (
            f"BaseIOConfig(\ninfile='{self.infile}'\noutfile='{self.outfile}'\n"
            f"file_type='{self.file_type}'\nsubsample_ratio={self.subsample_ratio}\n"
            f"level='{self.level}'\ncontext='{self.context}'\nprefix='{self.prefix}')"
        )

    def __repr__(self):
        return (
            f"BaseIOConfig(infile='{self.infile}', outfile='{self.outfile}', "
            f"subsample_ratio={self.subsample_ratio}, level='{self.level}', "
            f"context='{self.context}')"
        )

    def copy(self):
        """
        Return a deep copy of the config object.
        """
        return copy.deepcopy(self)
