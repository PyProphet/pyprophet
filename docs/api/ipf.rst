Inference of Peptidoforms Documentation
==========================

.. automodule:: pyprophet.ipf
   :no-members:
   :no-inherited-members:

.. currentmodule:: pyprophet.ipf

.. autosummary::
   :nosignatures:
   :toctree: generated/
   :template: class.rst

   infer_peptidoforms
   peptidoform_inference
   precursor_inference
   apply_bm
   prepare_transition_bm
   transfer_confident_evidence_across_runs
   prepare_precursor_bm
   compute_model_fdr