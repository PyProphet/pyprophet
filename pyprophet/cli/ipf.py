import click
from loguru import logger

from .util import write_logfile, measure_memory_usage_and_time
from .._config import IPFIOConfig
from ..ipf import infer_peptidoforms
from ..glyco.glycoform import infer_glycoforms


# IPF
@click.command()
# File handling
@click.option(
    "--in",
    "infile",
    required=True,
    type=click.Path(exists=True),
    help="PyProphet input file. Valid formats are .osw, .parquet (produced by export parquet)",
)
@click.option(
    "--out",
    "outfile",
    type=click.Path(exists=False),
    help="PyProphet output file. Valid formats are .osw, .parquet. Must be the same format as input file.",
)
# IPF parameters
@click.option(
    "--ipf_ms1_scoring/--no-ipf_ms1_scoring",
    default=True,
    show_default=True,
    help="Use MS1 precursor data for IPF.",
)
@click.option(
    "--ipf_ms2_scoring/--no-ipf_ms2_scoring",
    default=True,
    show_default=True,
    help="Use MS2 precursor data for IPF.",
)
@click.option(
    "--ipf_h0/--no-ipf_h0",
    default=True,
    show_default=True,
    help="Include possibility that peak groups are not covered by peptidoform space.",
)
@click.option(
    "--ipf_grouped_fdr/--no-ipf_grouped_fdr",
    default=False,
    show_default=True,
    help="[Experimental] Compute grouped FDR instead of pooled FDR to better support data where peak groups are evaluated to originate from very heterogeneous numbers of peptidoforms.",
)
@click.option(
    "--ipf_max_precursor_pep",
    default=0.7,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored precursors in IPF.",
)
@click.option(
    "--ipf_max_peakgroup_pep",
    default=0.7,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored peak groups in IPF.",
)
@click.option(
    "--ipf_max_precursor_peakgroup_pep",
    default=0.4,
    show_default=True,
    type=float,
    help="Maximum BHM layer 1 integrated precursor peakgroup PEP to consider in IPF.",
)
@click.option(
    "--ipf_max_transition_pep",
    default=0.6,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored transitions in IPF.",
)
@click.option(
    "--propagate_signal_across_runs/--no-propagate_signal_across_runs",
    default=False,
    show_default=True,
    help="Propagate signal across runs (requires running alignment).",
)
@click.option(
    "--ipf_max_alignment_pep",
    default=1.0,
    show_default=True,
    type=float,
    help="Maximum PEP to consider for good alignments.",
)
@click.option(
    "--across_run_confidence_threshold",
    default=0.5,
    show_default=True,
    type=float,
    help="Maximum PEP to consider for propagating signal across runs for aligned features.",
)
@click.pass_context
@measure_memory_usage_and_time
@logger.catch(reraise=True)
def ipf(
    ctx,
    infile,
    outfile,
    ipf_ms1_scoring,
    ipf_ms2_scoring,
    ipf_h0,
    ipf_grouped_fdr,
    ipf_max_precursor_pep,
    ipf_max_peakgroup_pep,
    ipf_max_precursor_peakgroup_pep,
    ipf_max_transition_pep,
    propagate_signal_across_runs,
    ipf_max_alignment_pep,
    across_run_confidence_threshold,
):
    """
    Infer peptidoforms after scoring of MS1, MS2 and transition-level data.
    """

    # Check to see if we're calling from the previous deprecated command call
    if ctx.command.deprecated:
        logger.warning(
            "The 'pyprophet ipf' command is deprecated and will be removed in future versions. Use 'pyprophet infer peptidoform' instead."
        )

    if outfile is None:
        outfile = infile
    else:
        outfile = outfile

    config = IPFIOConfig.from_cli_args(
        infile,
        outfile,
        1,  # Subsample ratio is not applicable for IPF
        "ipf",  # Level is not applicable for IPF
        "ipf",
        ipf_ms1_scoring,
        ipf_ms2_scoring,
        ipf_h0,
        ipf_grouped_fdr,
        ipf_max_precursor_pep,
        ipf_max_peakgroup_pep,
        ipf_max_precursor_peakgroup_pep,
        ipf_max_transition_pep,
        propagate_signal_across_runs,
        ipf_max_alignment_pep,
        across_run_confidence_threshold,
    )
    write_logfile(
        ctx.obj["LOG_LEVEL"], f"{config.prefix}_pyp_ipf.log", ctx.obj["LOG_HEADER"]
    )
    infer_peptidoforms(config)


# Infer glycoforms
@click.command()
@click.option(
    "--in", "infile", required=True, type=click.Path(exists=True), help="Input file."
)
@click.option("--out", "outfile", type=click.Path(exists=False), help="Output file.")
@click.option(
    "--ms1_precursor_scoring/--no-ms1_precursor_scoring",
    default=True,
    show_default=True,
    help="Use MS1 precursor data for glycoform inference.",
)
@click.option(
    "--ms2_precursor_scoring/--no-ms2_precursor_scoring",
    default=True,
    show_default=True,
    help="Use MS2 precursor data for glycoform inference.",
)
@click.option(
    "--grouped_fdr/--no-grouped_fdr",
    default=False,
    show_default=True,
    help="[Experimental] Compute grouped FDR instead of pooled FDR to better support data where peak groups are evaluated to originate from very heterogeneous numbers of glycoforms.",
)
@click.option(
    "--max_precursor_pep",
    default=1,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored precursors.",
)
@click.option(
    "--max_peakgroup_pep",
    default=0.7,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored peak groups.",
)
@click.option(
    "--max_precursor_peakgroup_pep",
    default=1,
    show_default=True,
    type=float,
    help="Maximum BHM layer 1 integrated precursor peakgroup PEP to consider.",
)
@click.option(
    "--max_transition_pep",
    default=0.6,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored transitions.",
)
@click.option(
    "--use_glycan_composition/--use_glycan_struct",
    "use_glycan_composition",
    default=True,
    show_default=True,
    help="Compute glycoform-level FDR based on glycan composition or struct.",
)
@click.option(
    "--ms1_mz_window",
    default=10,
    show_default=True,
    type=float,
    help="MS1 m/z window in Thomson or ppm.",
)
@click.option(
    "--ms1_mz_window_unit",
    default="ppm",
    show_default=True,
    type=click.Choice(["ppm", "Da", "Th"]),
    help="MS1 m/z window unit.",
)
@click.option(
    "--propagate_signal_across_runs/--no-propagate_signal_across_runs",
    default=False,
    show_default=True,
    help="Propagate signal across runs (requires running alignment).",
)
@click.option(
    "--max_alignment_pep",
    default=1.0,
    show_default=True,
    type=float,
    help="Maximum PEP to consider for good alignments.",
)
@click.option(
    "--across_run_confidence_threshold",
    default=0.5,
    show_default=True,
    type=float,
    help="Maximum PEP to consider for propagating signal across runs for aligned features.",
)
def glycoform(
    infile,
    outfile,
    ms1_precursor_scoring,
    ms2_precursor_scoring,
    grouped_fdr,
    max_precursor_pep,
    max_peakgroup_pep,
    max_precursor_peakgroup_pep,
    max_transition_pep,
    use_glycan_composition,
    ms1_mz_window,
    ms1_mz_window_unit,
    propagate_signal_across_runs,
    max_alignment_pep,
    across_run_confidence_threshold,
):
    """
    Infer glycoforms after scoring of MS1, MS2 and transition-level data.
    """

    if outfile is None:
        outfile = infile

    infer_glycoforms(
        infile=infile,
        outfile=outfile,
        ms1_precursor_scoring=ms1_precursor_scoring,
        ms2_precursor_scoring=ms2_precursor_scoring,
        grouped_fdr=grouped_fdr,
        max_precursor_pep=max_precursor_pep,
        max_peakgroup_pep=max_peakgroup_pep,
        max_precursor_peakgroup_pep=max_precursor_peakgroup_pep,
        max_transition_pep=max_transition_pep,
        use_glycan_composition=use_glycan_composition,
        ms1_mz_window=ms1_mz_window,
        ms1_mz_window_unit=ms1_mz_window_unit,
        propagate_signal_across_runs=propagate_signal_across_runs,
        max_alignment_pep=max_alignment_pep,
        across_run_confidence_threshold=across_run_confidence_threshold,
    )
