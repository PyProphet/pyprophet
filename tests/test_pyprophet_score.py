from __future__ import print_function

import os
import subprocess
import shutil
import sys

import pandas as pd
import sqlite3

import pytest

from pyprophet.ipf import read_pyp_peakgroup_precursor

pd.options.display.expand_frame_repr = False
pd.options.display.precision = 4
pd.options.display.max_columns = None

DATA_FOLDER = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data")

def _run_cmdline(cmdline):
    stdout = cmdline + "\n"
    try:
        stdout += str(subprocess.check_output(cmdline, shell=True,
                                          stderr=subprocess.STDOUT))
    except subprocess.CalledProcessError as error:
        print(error, end="", file=sys.stderr)
        raise
    return stdout

def _run_pyprophet_tsv_to_learn_model(regtest, temp_folder, dump_result_files=False, parametric=False, pfdr=False, pi0_lambda=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.txt")
    shutil.copy(data_path, temp_folder)
    cmdline = "pyprophet score --pi0_method=smoother --pi0_lambda 0.4 0 0 --in=test_data.txt --test --ss_iteration_fdr=0.02"
    if parametric:
        cmdline += " --parametric"
    if pfdr:
        cmdline += " --pfdr"
    if pi0_lambda is not False:
        cmdline += " --pi0_lambda=" + pi0_lambda
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_summary_stat.csv", sep=",", nrows=100).sort_index(axis=1),file=regtest)
    print(pd.read_csv("test_data_full_stat.csv", sep=",", nrows=100).sort_index(axis=1),file=regtest)
    print(pd.read_csv("test_data_scored.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    print(pd.read_csv("test_data_weights.csv", sep=",", nrows=100).sort_index(axis=1),file=regtest)

def _run_pyprophet_osw_to_learn_model_metabo(regtest, temp_folder, dump_result_files=False, ms1ms2=False, score_filter=False):
    os.chdir(temp_folder)
    data_path_metabo = os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path_metabo, temp_folder)

    # MS1-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms1 --test --ss_iteration_fdr=0.02"
    if score_filter:
        cmdline += " --ss_score_filter var_isotope_overlap_score,var_massdev_score,var_xcorr_coelution,var_isotope_correlation_score"

    # MS2-level
    if ms1ms2:
            cmdline += " score --in=test_data_compound.osw --level=ms1ms2 --test --ss_iteration_fdr=0.02"
            if score_filter:
                cmdline += " --ss_score_filter metabolomics"
    else:
        cmdline += " score --in=test_data_compound.osw --level=ms2 --test --ss_iteration_fdr=0.02"
        if score_filter:
            cmdline += " --ss_score_filter var_isotope_overlap_score,var_isotope_correlation_score,var_intensity_score,var_massdev_score,var_library_corr,var_norm_rt_score"

    stdout = _run_cmdline(cmdline)

    scored = sqlite3.connect("test_data_compound.osw")
    table = pd.read_sql_query("SELECT * FROM PYPROPHET_WEIGHTS ORDER BY score", scored)
    scored.close()

    print(table.head(100).sort_index(axis=1),file=regtest)

def _run_pyprophet_osw_to_learn_model(regtest, temp_folder, dump_result_files=False, parametric=False, pfdr=False, pi0_lambda=False, ms1ms2=False, xgboost=False, xgboost_tune=False, score_filter=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)

    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms1 --test --ss_iteration_fdr=0.02"
    if parametric:
        cmdline += " --parametric"
    if pfdr:
        cmdline += " --pfdr"
    if pi0_lambda is not False:
        cmdline += " --pi0_lambda=" + pi0_lambda
    if xgboost:
        cmdline += " --classifier=XGBoost"
    if xgboost_tune:
        cmdline += " --xgb_autotune"
    if score_filter:
        cmdline += " --ss_score_filter var_isotope_overlap_score,var_massdev_score,var_xcorr_coelution,var_isotope_correlation_score"

    # MS2-level
    if ms1ms2:
        cmdline += " score --in=test_data.osw --level=ms1ms2 --test --ss_iteration_fdr=0.02"
        if score_filter:
            cmdline += " --ss_score_filter var_ms1_isotope_overlap_score,var_ms1_massdev_score,var_ms1_xcorr_coelution,var_ms1_isotope_correlation_score,var_isotope_overlap_score,var_isotope_correlation_score,var_intensity_score,var_massdev_score,var_library_corr,var_norm_rt_score"
    else:
        cmdline += " score --in=test_data.osw --level=ms2 --test --ss_iteration_fdr=0.02"
        if score_filter:
            cmdline += " --ss_score_filter var_isotope_overlap_score,var_isotope_correlation_score,var_intensity_score,var_massdev_score,var_library_corr,var_norm_rt_score"
    if parametric:
        cmdline += " --parametric"
    if pfdr:
        cmdline += " --pfdr"
    if pi0_lambda is not False:
        cmdline += " --pi0_lambda=" + pi0_lambda
    if xgboost:
        cmdline += " --classifier=XGBoost"
    if xgboost_tune:
        cmdline += " --xgb_autotune"

    # transition-level
    cmdline += " score --in=test_data.osw --level=transition --test --ss_iteration_fdr=0.02"

    if parametric:
        cmdline += " --parametric"
    if pfdr:
        cmdline += " --pfdr"
    if pi0_lambda is not False:
        cmdline += " --pi0_lambda=" + pi0_lambda
    if xgboost:
        cmdline += " --classifier=XGBoost"
    if xgboost_tune:
        cmdline += " --xgb_autotune"

    stdout = _run_cmdline(cmdline)

    table = read_pyp_peakgroup_precursor("test_data.osw", 1.0, True, True)

    print(table.head(100).sort_index(axis=1),file=regtest)

def test_tsv_0(tmpdir, regtest):
    _run_pyprophet_tsv_to_learn_model(regtest, tmpdir.strpath, True)

def test_tsv_1(tmpdir, regtest):
    _run_pyprophet_tsv_to_learn_model(regtest, tmpdir.strpath, True, True)

def test_tsv_2(tmpdir, regtest):
    _run_pyprophet_tsv_to_learn_model(regtest, tmpdir.strpath, True, False, True)

def test_tsv_3(tmpdir, regtest):
    _run_pyprophet_tsv_to_learn_model(regtest, tmpdir.strpath, True, False, False, "0.3 0.55 0.05")

def test_tsv_apply_weights(tmpdir, regtest):

    _run_pyprophet_tsv_to_learn_model(regtest, tmpdir.strpath, True)

    _run_cmdline("pyprophet score --pi0_method=smoother --pi0_lambda 0.4 0 0 --in=test_data.txt --apply_weights=test_data_weights.csv "
                          "--test --ss_iteration_fdr=0.02")

def test_osw_0(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, pi0_lambda="0 0 0")

def test_osw_1(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, True, pi0_lambda="0 0 0")

def test_osw_2(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0")

def test_osw_3(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0", ms1ms2=True)

def test_osw_4(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0", ms1ms2=True, xgboost=True)

def test_osw_5(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0", ms1ms2=True, xgboost=True, xgboost_tune=True)

def test_osw_6(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0", score_filter=True)

def test_osw_7(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0", ms1ms2=True, score_filter=True)

def test_osw_8(tmpdir, regtest):
  _run_pyprophet_osw_to_learn_model_metabo(regtest, tmpdir.strpath, True, ms1ms2=False, score_filter=False)

def test_osw_9(tmpdir, regtest):
  _run_pyprophet_osw_to_learn_model_metabo(regtest, tmpdir.strpath, True, ms1ms2=False, score_filter=True)

def test_osw_10(tmpdir,regtest):
  _run_pyprophet_osw_to_learn_model_metabo(regtest, tmpdir.strpath, True, ms1ms2=True, score_filter=True)

def test_not_unique_tg_id_blocks(tmpdir):

    os.chdir(tmpdir.strpath)
    data_path = os.path.join(DATA_FOLDER, "test_invalid_data.txt")
    shutil.copy(data_path, tmpdir.strpath)
    cmdline = "pyprophet score --pi0_method=smoother --pi0_lambda 0.4 0 0 --in=test_invalid_data.txt --test --ss_iteration_fdr=0.02"

    with pytest.raises(subprocess.CalledProcessError) as exc_info:
        subprocess.check_output(cmdline, shell=True, stderr=subprocess.STDOUT)

    e = exc_info.value
    assert "Error: group_id values do not form unique blocks in input file(s)." in str(e.output)
