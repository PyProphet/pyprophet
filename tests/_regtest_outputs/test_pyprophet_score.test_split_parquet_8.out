96259
             feature_id  ms1_precursor_pep  ms2_peakgroup_pep  ms2_precursor_pep
0  -9078977811506172301             0.0064             0.0031             0.0987
1  -9059007664292712863             1.0000             0.3623                NaN
2  -9009602369958523731             0.0064             0.0031             0.4406
3  -8990894093332793487             0.0064             0.0031             0.0461
4  -8915955323477460297             0.0064             0.0031             0.0089
..                  ...                ...                ...                ...
95 -4539808410625597778             0.0064             0.0031             0.0461
96 -4495976808403190115             0.0064             0.0031             0.0227
97 -4474179539802460946             0.0064             0.0031             0.0089
98 -4409520928686189639             0.0064             0.0031             0.1061
99 -4399716566495748799             0.0064             0.0031                NaN

[100 rows x 4 columns]
96259
             feature_id  ms1_precursor_pep  ms2_peakgroup_pep  ms2_precursor_pep
0  -9078977811506172301             0.0064             0.0031             0.0987
1  -9059007664292712863             1.0000             0.3623                NaN
2  -9009602369958523731             0.0064             0.0031             0.4406
3  -8990894093332793487             0.0064             0.0031             0.0461
4  -8915955323477460297             0.0064             0.0031             0.0089
..                  ...                ...                ...                ...
95 -4539808410625597778             0.0064             0.0031             0.0461
96 -4495976808403190115             0.0064             0.0031             0.0227
97 -4474179539802460946             0.0064             0.0031             0.0089
98 -4409520928686189639             0.0064             0.0031             0.1061
99 -4399716566495748799             0.0064             0.0031                NaN

[100 rows x 4 columns]
